### PR TITLE
error handling for closed connections

### PR DESCRIPTION
**Pull request by blampe**
* cancel in-flight requests if our connection dies
* catch socket errors in `await`

We've still got a problem with callbacks in `await` where we hit an infinite recursion limit, looking into that.

### CODE REVIEW

File: python/tchannel/socket.py [hunk 1]
```diff
@@ -85,7 +99,12 @@ def outgoing(cls, hostport):
         conn.await_handshake_reply()
         return conn
 
+    @classmethod
+    def incoming(cls, hostport, handler):
+        raise NotImplementedError
```
Comment (abhinav) [inline python/tchannel/socket.py]: `()`

File: python/tchannel/socket.py [hunk 2]
```diff
@@ -62,6 +71,11 @@ def await(self):
             ctx = next(self.reader)
         except StopIteration:
             ctx = None
+        except socket.error as e:
```
Comment (abhinav) [inline python/tchannel/socket.py]: This will also catch timeouts. Shouldn't they have a different behavior?

Comment (blampe) [inline python/tchannel/socket.py]: yes but just trying to catch these for now

File: python/tchannel/tornado/connection.py [hunk 3]
```diff
@@ -64,28 +73,58 @@ def extract_handshake_headers(self, message):
...
+        # TODO: track awaiting responses in here
         message_id = message_id or self.next_message_id()
+
+        if message.message_type in (
+            messages.Types.CALL_REQ,
+            messages.Types.INIT_REQ,
+            messages.Types.PING_REQ,
+        ):
+            log.debug("awaiting response for message %s", message_id)
+            self.awaiting_responses[message_id] = gen.Future()
```
Comment (abhinav) [inline python/tchannel/tornado/connection.py]: Where is this future consumed?

Comment (blampe) [inline python/tchannel/tornado/connection.py]: In `send`.

**Comment by abhinav** [comment]
:shipit:

**Comment by breerly** [comment]
:shipit:

**Pull request merged**